### PR TITLE
Swap Notifier::Ses to use `aws-sdk` gem and add support for IAM instance profiles

### LINKED ISSUE #666: AWS:SES IAM Option
Hi,

Is it possible at the moment to use an IAM profile to authenticate the AWS:SES notifier? I'm using IAM for the s3 storage (`s3.use_iam_profile = true`) and would like to use this for SES notifications as well without having to specify the access key and secret in the backup model. Something like `ses.use_iam_profile = true`.

Thanks,
D

### PR DESCRIPTION
**Pull request by reedloden**
Sadly, the `aws-ses` gem doesn't support IAM instance profiles, so just swap it out for the official `aws-sdk` gem.

Fixes #666.

### CODE REVIEW

**Comment by reedloden** [comment]
See https://github.com/aws/aws-sdk-rails/blob/master/lib/aws/rails/mailer.rb for where most of this change comes from.

**Comment by reedloden** [comment]
@tombruijn -- I know you're looking for new maintainers and such, but in the mean time, anything I can do to help get this reviewed and landed? :)

**Comment by stuartellis** [comment]
@reedloden - From my perspective, the problem right now is getting integration tests into good enough shape to have the confidence to take changes like this.

If you are interested enough to help with AWS tests, that would be absolutely awesome. Otherwise, I'm afraid, it may be a while.

**Comment by reedloden** [comment]
@stuartellis - I'm all for good integration tests, but as far as I can tell, there are no integration tests really at all right now for anything (not just AWS-related things)... Is all development on `backup` blocked until there are viable integration tests? Or is there something else I can do to help test it manually enough for your satisfaction?

I added a few more tests, including using the [Client Response Stubs](https://aws.amazon.com/blogs/developer/client-response-stubs/) built into `aws-sdk`. The upstream `aws-sdk-rails` gem uses the stubs as well for testing: https://github.com/aws/aws-sdk-rails/blob/master/test/aws/rails/mailer_test.rb.

Funny enough, adding these additional tests caused me to discover a bug in the existing code related to BCC fields that comes out of hiding due to sending raw mail output to SES (rather than going through the `aws-ses` middleware that did its own black magic).

This is actively blocking me (working on a project to get rid of all hardcoded IAM credentials and move to IAM instance profiles), and I would prefer not to deploy a local fork if I don't need to. :)

File: lib/backup/notifier/mail.rb [hunk 1]
```diff
@@ -230,17 +230,3 @@ def new_email
...
-module Mail
-  class Exim
-    def self.call(path, arguments, _destinations, encoded_message)
-      popen "#{path} #{arguments}" do |io|
-        io.puts encoded_message.to_lf
-        io.flush
-      end
-    end
-  end
-end
```
Comment (tombruijn) [inline lib/backup/notifier/mail.rb]: Have you tested removing the monkeypatch doesn't break exim support?

File: spec/notifier/ses_spec.rb [hunk 2]
```diff
@@ -1,4 +1,5 @@
 require "spec_helper"
+require "aws-sdk"
```
Comment (tombruijn) [inline spec/notifier/ses_spec.rb]: Should be automatically required by bundler, right?

File: lib/backup/notifier/mail.rb (hunk 1 above)
Comment (reedloden) [inline lib/backup/notifier/mail.rb]: I upgraded `mail` to v2.5.4, which includes the monkey patch, so it's not needed.

File: spec/notifier/ses_spec.rb (hunk 2 above)
Comment (reedloden) [inline spec/notifier/ses_spec.rb]: Indeed, not needed as far as I can tell. I removed it.

File: lib/backup/notifier/mail.rb (hunk 1 above)
Comment (reedloden) [inline lib/backup/notifier/mail.rb]: Specifically, see https://github.com/backup/backup/issues/446 and https://github.com/mikel/mail/pull/546. Don't need the monkey patch anymore with the `mail` version bump. :)

**Comment by reedloden** [comment]
@tombruijn, I can split it into two PRs if it would help... Up to you. Whatever helps get it landed faster. :)

File: lib/backup/notifier/mail.rb (hunk 1 above)
Comment (tombruijn) [inline lib/backup/notifier/mail.rb]: Not sure what you mean. If I look at version 2.6.4 of mail I still see the same call to `#encoded`: https://github.com/mikel/mail/compare/2.6.3...2.6.4?expand=1#diff-70e36d3e2b37ee7aca19125cd388fd6fR47
The PR made by an earlier backup maintainer wasn't merged.

Comment (reedloden) [inline lib/backup/notifier/mail.rb]: Ugh, my bad. I somehow had it in my head that it had been merged and wasn't needed. My apologies! I've reverted that change. This should be good to go now.

File: backup.gemspec [hunk 3]
```diff
@@ -36,13 +36,13 @@ Gem::Specification.new do |gem|
   gem.add_dependency "net-ssh", "3.2.0"
   gem.add_dependency "net-scp", "1.2.1"
   gem.add_dependency "net-sftp", "2.1.2"
-  gem.add_dependency "mail", "2.6.3" # patched
+  gem.add_dependency "mail", "2.6.4" # patched
```
Comment (tombruijn) [inline backup.gemspec]: The upgrade to this means we do need to update the monkeypatch to be like this:
source: https://github.com/mikel/mail/compare/2.6.3...2.6.4#diff-70e36d3e2b37ee7aca19125cd388fd6fR47
`lib/mail/network/delivery_methods/exim.rb`

```ruby
# lib/backup/notifier/mail.rb:241
io.puts ::Mail::Utilities.to_lf(encoded_mail)
```

File: lib/backup/notifier/ses.rb [hunk 4]
```diff
@@ -96,7 +106,15 @@ def notify!(status)
           }
         end
 
-        client.send_raw_email(email)
+        send_opts = {}
+        send_opts[:raw_message] = {}
+        send_opts[:raw_message][:data] = email.to_s
```
Comment (tombruijn) [inline lib/backup/notifier/ses.rb]: if you're not merging `send_opts` with another hash, can you simply this?

```ruby
send_opts = {
  raw_message: {
    data: email.to_s
  }
}
```

File: lib/backup/notifier/ses.rb [hunk 5]
```diff
@@ -1,11 +1,12 @@
...
+require "aws-sdk"
+require "mail"
 
 module Backup
   module Notifier
     class Ses < Base
       ##
       # Amazon Simple Email Service (SES) Credentials
-      attr_accessor :access_key_id, :secret_access_key
+      attr_accessor :access_key_id, :secret_access_key, :use_iam_profile
```
Comment (tombruijn) [inline lib/backup/notifier/ses.rb]: can you add test scenarios that use this option?

File: backup.gemspec (hunk 3 above)
Comment (reedloden) [inline backup.gemspec]: Changed, though assume you mean `encoded_message` :)

File: lib/backup/notifier/ses.rb (hunk 4 above)
Comment (reedloden) [inline lib/backup/notifier/ses.rb]: I had purposely kept it like this, as it's copy/paste from `aws-sdk-rails` (https://github.com/aws/aws-sdk-rails/blob/9c1487dc58a0d42b5587f4749cbc10442d6aa023/lib/aws/rails/mailer.rb#L26-L28), but happy to change.

File: lib/backup/notifier/ses.rb (hunk 5 above)
Comment (reedloden) [inline lib/backup/notifier/ses.rb]: Done! I think I did `shared_examples` correctly to DRY it up somewhat...

File: backup.gemspec (hunk 3 above)
Comment (tombruijn) [inline backup.gemspec]: I did yes :)

**Comment by tombruijn** [comment]
I've added a commit that changes how we test the credentials, by mocking on a bit higher level than stubbing the client method in the notifier itself. Now I think it's 👍  
@stuartellis what do you think?

**Comment by stuartellis** [comment]
@tombruijn - I don't feel too qualified to comment :). 

@reedloden: Since it's just mocks, it would be great if you could confirm that the final PR works for reals with SES and IAM.

FWIW, I see this warning when running with Ruby 2.4.1:

https://github.com/mikel/mail/issues/1068

The fix is in mail 2.6.5.

**Comment by tombruijn** [comment]
> @reedloden: Since it's just mocks, it would be great if you could confirm that the final PR works for reals with SES and IAM.

agreed

> FWIW, I see this warning when running with Ruby 2.4.1: mikel/mail#1068 The fix is in mail 2.6.5.

It's not crucial, but if you could upgrade this @reedloden that would be great :)

**Comment by reedloden** [comment]
@stuartellis / @tombruijn -- I bumped `mail` to v2.6.5 and did another local test-run with this branch, and I can confirm that IAM instance profiles do indeed work fine. :)

:shipit:

**Pull request merged**